File: python/string/rolling_hash.py
# 1-dimension Rolling Hash
class RollingHash:
    def __init__(self, s, base, MOD):
        self.s = s
        self.l = l = len(s)
        self.base = base
        self.MOD = MOD
        self.h = h = [0]*(l + 1)
        for i in range(l):
            h[i+1] = (h[i] * base + ord(s[i])) % MOD
    def get(self, l, r):
        MOD = self.MOD
        return (self.h[r] - self.h[l]*pow(self.base, r-l, MOD)) % MOD

# 非クラス版
def rolling_hash(s, base, MOD):
    l = len(s)
    h = [0]*(l + 1)
    for i in range(l):
        h[i+1] = (h[i] * base + ord(s[i])) % MOD
    return h
def get(h, l, r, base, MOD):
    return (h[r] - h[l]*pow(base, r-l, MOD)) % MOD

File: cpp/string/rolling_hash.cpp
#include<string>
using namespace std;
using ll = long long;

#define L 1000004

class RollingHash {
  ll base, mod;
  ll ptable[L];
  ll h[L];
  int sz;
public:
  RollingHash(string s, ll base, ll mod) : base(base), mod(mod) {
    sz = s.length();
    h[0] = 0;
    for(int i=0; i<sz; ++i) {
      h[i+1] = (h[i] * base + s[i]) % mod;
    }
    ptable[0] = 1;
    for(int i=0; i<sz; ++i) {
      ptable[i+1] = (ptable[i] * base) % mod;
    }
  }

  ll get(int l, int r) {
    // assert(0 <= l && r <= sz && l <= r);
    return (h[r] + ((mod - h[l]) * ptable[r-l] % mod)) % mod;
  }
};
